r2-core/src/test/java/test/r2/transport/http/client/TestAsyncPool.java

```diff
@@ -570,6 +580,389 @@ public void testWaitTimeStats() throws Exception
+   *       objects even though the waiters are timedout.
+   *
+   * @param numberOfCheckoutsInPhaseA the N number of checkout operations that will be performed in phase A
+   * @param numberOfCheckoutsInPhaseB the O number of checkout operations that will be performed in Phase B
+   * @param numbOfObjectsToBeReturnedInPhaseC the numeber of objects returned in Phase C
+   * @param poolSize size of the pool,
+   * @param concurrency concurrency of the rate limiter
+   */
+  @Test(dataProvider = "waiterTimeoutDataProvider")
+  public void testWaiterTimeout(int numberOfCheckoutsInPhaseA, int numberOfCheckoutsInPhaseB,
+      int numbOfObjectsToBeReturnedInPhaseC,
+      int poolSize, int concurrency, int waiterTimeout)
+  {
+    CreationBlockableSynchronousLifecycle blockableObjectCreator =
+        new CreationBlockableSynchronousLifecycle(numberOfCheckoutsInPhaseB, concurrency);
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(500);
+    RateLimiter rateLimiter = new ExponentialBackOffRateLimiter(0, 5000,
+        10, executor, concurrency);
+
+    ClockedExecutor clockedExecutor = new ClockedExecutor();
+    SettableClock settableClock = new SettableClock();
+
+    final AsyncPool<Object> pool = new AsyncPoolImpl<Object>("object pool",
+        blockableObjectCreator,
+        poolSize,
+        Integer.MAX_VALUE,
+        waiterTimeout,
+        clockedExecutor,
+        Integer.MAX_VALUE,
+        AsyncPoolImpl.Strategy.MRU,
+        0, rateLimiter, settableClock, new LongTracking()
+    );
+
+    pool.start();
+
+    // Phase A : Checking out object 'numberOfCheckoutsInPhaseA' times !
+    List<Object> checkedOutObjects = performCheckout(numberOfCheckoutsInPhaseA, pool);
+
+    // Phase B : Blocking object creation and performing the checkout 'numberOfCheckoutsInPhaseB' times again
+    blockableObjectCreator.blockCreation();
+    Future<None> future = performUnblockingCheckout(numberOfCheckoutsInPhaseB,
+        0, pool);
+
+    blockableObjectCreator.waitUntilAllBlocked();
+
+    // Phase C : Returning the checkedOut objects from Phase A back to the object pool
+    for (int i = 0; i < numbOfObjectsToBeReturnedInPhaseC; i++)
+    {
+      pool.put(checkedOutObjects.remove(0));
+    }
+
+    settableClock.addDuration(waiterTimeout+1);
+    clockedExecutor.runFor(waiterTimeout+2);
+
+    // Phase D : All the object creation in phase B gets unblocked now
+    blockableObjectCreator.unblockCreation();
+    try
+    {
+      future.get(5, TimeUnit.SECONDS);
+    }
+    catch (Exception e)
+    {
+      Assert.fail("Unexpected interruption", e);
+    }
+
+    // Making sure the rate limiter pending tasks are submitted to the executor
+    while (rateLimiter.numberOfPendingTasks() > 0)
+    {
+      try
+      {
+        Thread.sleep(1);
+      }
+      catch (Exception e)
+      {
+        Assert.fail("Unexpected interruption", e);
+      }
+    }
```

Comment:
use AssertWithTimeout instead to avoid locking situations, here and in the other places with sleep

Reply:
Thanks.